File: java_version/trees/sumTree.java
class Solution {
    
    boolean isSumTree(Node root) {
        return helper(root) != -1;
    }
    
    int helper(Node node) {
        
        if (node == null) return 0;
        
        // leaf node
        if (node.left == null && node.right == null) {
            return node.data;
        }
        
        int left = helper(node.left);
        int right = helper(node.right);
        
        // if any subtree is invalid
        if (left == -1 || right == -1) return -1;
        
        // check sum tree condition
        if (node.data == left + right) {
            return node.data + left + right;
        } else {
            return -1;
        }
    }
}

File: trees/sumTree.py
class Solution:
    def is_sum_tree(self, node):
        
        def helper(node):
            if node is None:
                return (True, 0)
            
            if node.left is None and node.right is None:
                return (True, node.data)
            
            left_ok, left_sum = helper(node.left)
            right_ok, right_sum = helper(node.right)
            
            if node.data == left_sum + right_sum:
                return (left_ok and right_ok, node.data + left_sum + right_sum)
            else:
                return (False, 0)
        
        return helper(node)[0]
